### PR TITLE
Update version of Mbed TLS

### PR DESCRIPTION
**Pull request by egrimley-arm**
`third-party/rust-mbedtls` is now based on the upstream branch `v0.10`, tag `mbedtls_v0.10.0`.

(It was initially based on https://github.com/fortanix/rust-mbedtls/pull/278, which has now been merged.)

### CODE REVIEW

**Comment by mathias-arm** [comment]
The upgrade to `bindgen` version 0.65.1 brings additional dependency changes. It would be interesting to see if we can downgrade to 0.64 to avoid that version bump on `syn`. I think there is an issue with `psa-attestation` if we try to upgrade `bindgen` in that crate.

**Comment by egrimley-arm** [comment]
I can confirm that `mbedtls-sys` does not work with `bindgen` `0.64`, and `psa-attestation` does not work with `bindgen` `0.65.1`. (I haven't tried to make sense of the errors generated in either case.) The semantic versioning of Rust crates treats the left-most non-zero number as major, so we can't expect compatibility between these versions of `bindgen`, unfortunately. I hope Cargo keeps build-dependencies separate from ordinary dependencies so the damage is at least limited.

**Comment by egrimley-arm** [comment]
> Your new branch of mbedtls did not pick those changes:
> [veracruz-project/rust-mbedtls@7cac229](https://github.com/veracruz-project/rust-mbedtls/commit/7cac2294819ebce0859eb57d6aa94bb301820c2b)
> [veracruz-project/rust-mbedtls@0b62e75](https://github.com/veracruz-project/rust-mbedtls/commit/0b62e751612fa53017c9476a5bd65b9d2b67df00)

I've rescued those two: just the `Cargo.toml` changes as the `Cargo.lock` changes would get in the way of the merging and rebasing that I will have to do.

Are the following commits still of interest?

[veracruz-project/rust-mbedtls@7a72ec0](https://github.com/veracruz-project/rust-mbedtls/commit/7a72ec096a0f6693b34f734c6983b96a2ff4d526)
[veracruz-project/rust-mbedtls@394e7ae](https://github.com/veracruz-project/rust-mbedtls/commit/394e7ae59e5d6de57c420b80db724ed921da67d4)

**Comment by mathias-arm** [comment]
> > Your new branch of mbedtls did not pick those changes:
> > [veracruz-project/rust-mbedtls@7cac229](https://github.com/veracruz-project/rust-mbedtls/commit/7cac2294819ebce0859eb57d6aa94bb301820c2b)
> > [veracruz-project/rust-mbedtls@0b62e75](https://github.com/veracruz-project/rust-mbedtls/commit/0b62e751612fa53017c9476a5bd65b9d2b67df00)
> 
> I've rescued those two: just the `Cargo.toml` changes as the `Cargo.lock` changes would get in the way of the merging and rebasing that I will have to do.
> 
> Are the following commits still of interest?
> 
> [veracruz-project/rust-mbedtls@7a72ec0](https://github.com/veracruz-project/rust-mbedtls/commit/7a72ec096a0f6693b34f734c6983b96a2ff4d526) [veracruz-project/rust-mbedtls@394e7ae](https://github.com/veracruz-project/rust-mbedtls/commit/394e7ae59e5d6de57c420b80db724ed921da67d4)

We need to test on AArch64 to see if those are still needed.

**Comment by egrimley-arm** [comment]
Since the change made by [veracruz-project/rust-mbedtls@7cac229](https://github.com/veracruz-project/rust-mbedtls/commit/7cac2294819ebce0859eb57d6aa94bb301820c2b) is in a section called `[target.x86_64-fortanix-unknown-sgx.dependencies]`, presumably it makes no real difference to us, so the only purpose of the change is to keep Dependabot happy?

**Comment by mathias-arm** [comment]
Looking at the branch in [rust-mbedtls](https://github.com/veracruz-project/rust-mbedtls/commits/egrimley/pr-update-mbedtls) there is some back and forth (things done and then reverted), maybe some squashing would be helpful.

**Comment by egrimley-arm** [comment]
I've rebased, squashed and improved the log messages.

We could wait for the upstream branch to get merged or released or for development of the upstream branch to slow down before merging this, or we could not wait for that.

**Comment by egrimley-arm** [comment]
* We should merge your work into `main` branch of [rust-mbedtls](https://github.com/veracruz-project/rust-mbedtls).

We've been using the `veracruz` branch, though it's not mentioned in `veracruz/.gitmodules`: perhaps it should be and I could add that in this PR?

**Comment by mathias-arm** [comment]
> We've been using the `veracruz` branch, though it's not mentioned in `veracruz/.gitmodules`: perhaps it should be and I could add that in this PR?

Good idea. Based on today's discussion, I think we should not wait for an upstream merge and/or release.

**Comment by mathias-arm** [comment]
Can you remove the `Remove unused features "icecap-lkvm" and "icecap-qemu".` commit from the PR, this might create an issue for #555?

**Comment by egrimley-arm** [comment]
> Can you remove the `Remove unused features "icecap-lkvm" and "icecap-qemu".` commit from the PR, this might create an issue for #555?

OK. By the way, there seems to be an "approved" on the upstream PR (https://github.com/fortanix/rust-mbedtls/pull/278). I don't know if that means an imminent merge. I might rebase again on Monday.

**Review by mathias-arm — APPROVED** [review]

**Comment by egrimley-arm** [comment]
The upstream PR has just been merged into `master`, 49 minutes ago, but it's been rebased so the commit I'm basing this on is not in `master`. So I think it's worth me rebasing this again before we merge.

**Comment by egrimley-arm** [comment]
Done. Almost nothing changed: `git diff --submodule=diff 81be98c..e11872e`

@mathias-arm, if you can approve again I'll merge this time.

**Comment by egrimley-arm** [comment]
I've just rebased again.

**Review by dreemkiller — APPROVED** [review]

**Pull request merged**
